File: scripts/cloud/forensic-worker-driver.mjs
#!/usr/bin/env node

import {
  constants as fsConstants,
  accessSync,
  existsSync,
  lstatSync,
  readFileSync,
  readdirSync,
  readlinkSync,
  rmSync,
} from "node:fs";
import { spawnSync } from "node:child_process";
import { basename, join } from "node:path";
import { pathToFileURL } from "node:url";

const DRIVER_REF = "driver.openagents.forensic-worker.v1";
const BUBBLEWRAP = "/usr/bin/bwrap";
const WORKSPACE = "/workspace";
const TURN_ROOT = "/var/lib/openagents/managed-sandbox-turns";
const SOURCE_ROOT = `${WORKSPACE}/source`;
const ARTIFACT_PATH = `${WORKSPACE}/forensic-artifact.tar.zst`;
const RUNTIME_ROOT = "/run/openagents-managed-sandbox";
const NETWORK_ROOT = "/sys/class/net";

const refuse = (reason) => {
  process.stderr.write(`${JSON.stringify({ error: "forensic_worker_refused", reason })}\n`);
  process.exit(1);
};

const emit = (value) => process.stdout.write(`${JSON.stringify(value)}\n`);

const liveIdentifier = (path, processGroup) => {
  try {
    const identifier = Number.parseInt(readFileSync(path, "utf8"), 10);
    if (!Number.isSafeInteger(identifier) || identifier < 1) return true;
    process.kill(processGroup ? -identifier : identifier, 0);
    return true;
  } catch (error) {
    return error?.code !== "ESRCH";
  }
};

const directories = (turnRoot, includeIo) =>
  existsSync(turnRoot)
    ? readdirSync(turnRoot, { withFileTypes: true })
        .filter((entry) => entry.isDirectory() && (includeIo || !entry.name.startsWith("io-")))
        .map((entry) => join(turnRoot, entry.name))
    : [];

const turnDirectories = () => directories(TURN_ROOT, false);

export const workloadIsLiveAt = (directory) =>
  existsSync(join(directory, "pgid"))
    ? liveIdentifier(join(directory, "pgid"), true)
    : existsSync(join(directory, "pid"))
      ? liveIdentifier(join(directory, "pid"), false)
      : !basename(directory).startsWith("io-");

// A residue check must not follow symlinks: `existsSync` resolves them, so a
// dangling symlink left behind after removal would be reported as clean.
const pathPresent = (path) => {
  try {
    lstatSync(path);
    return true;
  } catch {
    return false;
  }
};

const pathIsUnder = (path, root) => path === root || path.startsWith(`${root}/`);

// Observe every live process that still references a guarded root through its
// working directory, filesystem root, executable, or any open descriptor. This
// is the only observation that survives a descendant escaping its process group
// with setsid()/setpgid(), so it is what makes "zero process" a measurement
// rather than an assertion.
export const observeGuardedProcessesAt = (roots) => {
  if (process.platform !== "linux" || !existsSync("/proc")) {
    return { supported: false, processes: [], processGroups: [] };
  }
  const processes = [];
  const processGroups = new Set();
  const self = process.pid;
  for (const entry of readdirSync("/proc", { withFileTypes: true })) {
    if (!entry.isDirectory() || !/^\d+$/u.test(entry.name)) continue;
    const pid = Number.parseInt(entry.name, 10);
    if (pid === self) continue;
    const references = [`/proc/${pid}/cwd`, `/proc/${pid}/root`, `/proc/${pid}/exe`];
    try {
      for (const descriptor of readdirSync(`/proc/${pid}/fd`)) {
        references.push(`/proc/${pid}/fd/${descriptor}`);
      }
    } catch (error) {
      // A process that exits mid-scan is not residue. Any other failure means
      // the observation is incomplete and must not be reported as clean.
      if (error?.code !== "ENOENT" && error?.code !== "ESRCH") {
        return { supported: false, processes: [], processGroups: [] };
      }
    }
    let referenced = false;
    for (const reference of references) {
      let target;
      try {
        target = readlinkSync(reference);
      } catch {
        continue;
      }
      if (roots.some((root) => pathIsUnder(target, root))) {
        referenced = true;
        break;
      }
    }
    if (!referenced) continue;
    processes.push(pid);
    try {
      const stat = readFileSync(`/proc/${pid}/stat`, "utf8");
      const close = stat.lastIndexOf(")");
      const fields = close < 0 ? [] : stat.slice(close + 2).split(" ");
      const processGroupId = Number.parseInt(fields[2] ?? "", 10);
      if (Number.isSafeInteger(processGroupId)) processGroups.add(processGroupId);
    } catch (error) {
      if (error?.code !== "ENOENT" && error?.code !== "ESRCH") {
        return { supported: false, processes: [], processGroups: [] };
      }
    }
  }
  return {
    supported: true,
    processes,
    processGroups: [...processGroups].toSorted((left, right) => left - right),
  };
};

const fileSize = (path) => {
  if (!existsSync(path)) return { bytes: 0, exact: true };
  try {
    const status = lstatSync(path);
    return status.isFile()
      ? { bytes: status.size, exact: Number.isSafeInteger(status.size) }
      : { bytes: 0, exact: false };
  } catch {
    return { bytes: 0, exact: false };
  }
};

const treeSize = (root) => {
  if (!existsSync(root)) return { bytes: 0, exact: true };
  let bytes = 0;
  let exact = true;
  const pending = [root];
  while (pending.length > 0) {
    const path = pending.pop();
    try {
      const status = lstatSync(path);
      if (status.isSymbolicLink()) exact = false;
      else if (status.isFile()) bytes += status.size;
      else if (status.isDirectory()) {
        for (const entry of readdirSync(path)) pending.push(join(path, entry));
      } else exact = false;
      if (!Number.isSafeInteger(bytes)) exact = false;
    } catch {
      exact = false;
    }
  }
  return { bytes, exact };
};

const networkUsage = () => {
  if (process.platform !== "linux" || !existsSync(NETWORK_ROOT)) {
    return { bytes: 0, exact: false };
  }
  let bytes = 0;
  let observed = false;
  try {
    for (const interfaceName of readdirSync(NETWORK_ROOT)) {
      if (interfaceName === "lo") continue;
      observed = true;
      for (const counter of ["rx_bytes", "tx_bytes"]) {
        const value = Number.parseInt(
          readFileSync(join(NETWORK_ROOT, interfaceName, "statistics", counter), "utf8"),
          10,
        );
        if (!Number.isSafeInteger(value) || value < 0) return { bytes: 0, exact: false };
        bytes += value;
        if (!Number.isSafeInteger(bytes)) return { bytes: 0, exact: false };
      }
    }
  } catch {
    return { bytes: 0, exact: false };
  }
  return { bytes, exact: observed };
};

const preflight = () => {
  if (process.platform !== "linux") refuse("linux_required");
  try {
    accessSync(BUBBLEWRAP, fsConstants.X_OK);
    accessSync(WORKSPACE, fsConstants.R_OK | fsConstants.X_OK);
    const probe = spawnSync(
      BUBBLEWRAP,
      [
        "--die-with-parent",
        "--unshare-net",
        "--unshare-pid",
        "--unshare-uts",
        "--unshare-ipc",
        "--ro-bind",
        "/",
        "/",
        "--bind",
        WORKSPACE,
        WORKSPACE,
        "--tmpfs",
        "/run",
        "--proc",
        "/proc",
        "--dev",
        "/dev",
        "--chdir",
        WORKSPACE,
        "/bin/true",
      ],
      {
        cwd: WORKSPACE,
        env: { PATH: "/usr/bin:/bin" },
        encoding: "utf8",
        timeout: 5_000,
        windowsHide: true,
      },
    );
    if (probe.error !== undefined || probe.status !== 0 || probe.signal !== null) {
      refuse("bubblewrap_probe_failed");
    }
  } catch {
    refuse("runtime_dependency_missing");
  }
  emit({
    schema: "openagents.forensic_worker_preflight.v1",
    driverRef: DRIVER_REF,
    linux: true,
    bubblewrap: true,
    networkNamespace: "unshared",
    workspaceRoot: "workspace",
  });
};

const usage = () => {
  let exact = true;
  let tokens = 0;
  let activeTurns = 0;
  for (const directory of turnDirectories()) {
    if (workloadIsLiveAt(directory)) activeTurns += 1;
    try {
      const state = JSON.parse(readFileSync(join(directory, "state.json"), "utf8"));
      const usageEvents = Array.isArray(state.events)
        ? state.events.filter((event) => event?._tag === "RuntimeUsageRecorded")
        : [];
      if (usageEvents.length !== 1) exact = false;
      for (const event of usageEvents) {
        if (event?.usage?.exact !== true) exact = false;
        const inputTokens = event?.usage?.inputTokens;
        const outputTokens = event?.usage?.outputTokens;
        if (!Number.isSafeInteger(inputTokens) || !Number.isSafeInteger(outputTokens))
          exact = false;
        else tokens += inputTokens + outputTokens;
      }
    } catch {
      exact = false;
    }
  }
  const source = treeSize(SOURCE_ROOT);
  const artifact = fileSize(ARTIFACT_PATH);
  const network = networkUsage();
  emit({
    exact: exact && source.exact && artifact.exact && network.exact,
    tokens,
    sourceBytes: source.bytes,
    artifactBytes: artifact.bytes,
    networkBytes: network.bytes,
    activeTurns,
  });
};

export const prepareStopAt = ({
  turnRoot,
  sourceRoot,
  artifactPath,
  runtimeRoot,
  observeGuardedProcesses = observeGuardedProcessesAt,
}) => {
  const guardedRoots = [turnRoot, sourceRoot, artifactPath, runtimeRoot];

  // The turn-directory scan only covers recorded pid/pgid files beneath
  // `turnRoot`. The guarded-root observation additionally covers the source,
  // artifact, and runtime roots, which were previously removed with no
  // liveness check at all.
  if (directories(turnRoot, true).some(workloadIsLiveAt)) {
    refuse("forensic_process_still_active");
  }
  const before = observeGuardedProcesses(guardedRoots);
  if (process.platform === "linux" && !before.supported) {
    refuse("forensic_process_observation_unavailable");
  }
  if (before.processes.length !== 0) {
    refuse("forensic_process_still_active");
  }

  rmSync(turnRoot, { recursive: true, force: true });
  rmSync(sourceRoot, { recursive: true, force: true });
  rmSync(artifactPath, { recursive: true, force: true });
  rmSync(runtimeRoot, { recursive: true, force: true });

  // Re-observe after removal. These counts are the proof; they are never
  // assumed from the fact that removal returned without throwing.
  const after = observeGuardedProcesses(guardedRoots);
  if (process.platform === "linux" && !after.supported) {
    refuse("forensic_process_observation_unavailable");
  }
  const scratchPathsRemaining = guardedRoots.filter(pathPresent).length;
  const activeProcessGroups = after.processGroups.length;
  if (after.processes.length !== 0) {
    refuse("forensic_process_still_active");
  }
  if (scratchPathsRemaining !== 0) {
    refuse("forensic_scratch_still_present");
  }
  return {
    schema: "openagents.forensic_worker_prepare_stop.v1",
    driverRef: DRIVER_REF,
    processObservation: after.supported ? "proc" : "unavailable",
    zeroProcess: after.processes.length === 0 && activeProcessGroups === 0,
    zeroScratch: scratchPathsRemaining === 0,
    activeProcessGroups,
    scratchPathsRemaining,
  };
};

const prepareStop = () =>
  emit(
    prepareStopAt({
      turnRoot: TURN_ROOT,
      sourceRoot: SOURCE_ROOT,
      artifactPath: ARTIFACT_PATH,
      runtimeRoot: RUNTIME_ROOT,
    }),
  );

if (process.argv[1] !== undefined && import.meta.url === pathToFileURL(process.argv[1]).href) {
  if (process.argv.length !== 3) refuse("unsupported_operation");
  if (process.argv[2] === "preflight") preflight();
  else if (process.argv[2] === "usage") usage();
  else if (process.argv[2] === "prepare-stop") prepareStop();
  else refuse("unsupported_operation");
}

File: scripts/cloud/forensic-worker-driver.py
#!/usr/bin/env python3
"""Guest forensic worker: preflight, usage, and prepare-stop proofs."""

from __future__ import annotations

import json
import os
import shutil
import stat
import subprocess
import sys
from pathlib import Path
from typing import Any, Callable


DRIVER_REF = "driver.openagents.forensic-worker.v1"
BUBBLEWRAP = "/usr/bin/bwrap"
WORKSPACE = "/workspace"
TURN_ROOT = "/var/lib/openagents/managed-sandbox-turns"
SOURCE_ROOT = f"{WORKSPACE}/source"
ARTIFACT_PATH = f"{WORKSPACE}/forensic-artifact.tar.zst"
RUNTIME_ROOT = "/run/openagents-managed-sandbox"
NETWORK_ROOT = "/sys/class/net"


def refuse(reason: str) -> None:
    sys.stderr.write(json.dumps({"error": "forensic_worker_refused", "reason": reason}) + "\n")
    raise SystemExit(1)


def emit(value: Any) -> None:
    sys.stdout.write(json.dumps(value) + "\n")


def live_identifier(path: Path, process_group: bool) -> bool:
    try:
        identifier = int(path.read_text(encoding="utf-8"))
        if identifier < 1:
            return True
        os.kill(-identifier if process_group else identifier, 0)
        return True
    except ProcessLookupError:
        return False
    except (OSError, ValueError):
        return True


def directories(turn_root: str, include_io: bool) -> list[str]:
    root = Path(turn_root)
    if not root.exists():
        return []
    found: list[str] = []
    for entry in root.iterdir():
        if not entry.is_dir():
            continue
        if not include_io and entry.name.startswith("io-"):
            continue
        found.append(str(entry))
    return found


def workload_is_live_at(directory: str) -> bool:
    pgid = Path(directory) / "pgid"
    pid = Path(directory) / "pid"
    if pgid.exists():
        return live_identifier(pgid, True)
    if pid.exists():
        return live_identifier(pid, False)
    return not Path(directory).name.startswith("io-")


def path_present(path: str) -> bool:
    try:
        os.lstat(path)
        return True
    except OSError:
        return False


def path_is_under(path: str, root: str) -> bool:
    return path == root or path.startswith(f"{root}/")


def observe_guarded_processes_at(roots: list[str]) -> dict[str, Any]:
    if sys.platform != "linux" or not Path("/proc").exists():
        return {"supported": False, "inaccessible": 0, "processes": [], "processGroups": []}
    processes: list[int] = []
    process_groups: set[int] = set()
    self_pid = os.getpid()
    inaccessible = 0
    for entry in Path("/proc").iterdir():
        if not entry.is_dir() or not entry.name.isdigit():
            continue
        pid = int(entry.name)
        if pid == self_pid:
            continue
        references = [f"/proc/{pid}/cwd", f"/proc/{pid}/root", f"/proc/{pid}/exe"]
        opaque = False
        try:
            for descriptor in Path(f"/proc/{pid}/fd").iterdir():
                references.append(str(descriptor))
        except OSError as error:
            if getattr(error, "errno", None) not in (2, 3):
                opaque = True
        referenced = False
        for reference in references:
            try:
                target = os.readlink(reference)
            except OSError as error:
                if getattr(error, "errno", None) not in (2, 3):
                    opaque = True
                continue
            if any(path_is_under(target, root) for root in roots):
                referenced = True
                break
        if referenced:
            processes.append(pid)
            try:
                stat_text = Path(f"/proc/{pid}/stat").read_text(encoding="utf-8")
                close = stat_text.rfind(")")
                fields = [] if close < 0 else stat_text[close + 2 :].split(" ")
                process_group_id = int(fields[2]) if len(fields) > 2 else None
                if process_group_id is None:
                    opaque = True
                else:
                    process_groups.add(process_group_id)
            except (OSError, ValueError, IndexError) as error:
                if getattr(error, "errno", None) not in (2, 3):
                    opaque = True
        if opaque and not referenced:
            inaccessible += 1
    return {
        "supported": True,
        "inaccessible": inaccessible,
        "processes": processes,
        "processGroups": sorted(process_groups),
    }


def file_size(path: str) -> dict[str, Any]:
    if not Path(path).exists():
        return {"bytes": 0, "exact": True}
    try:
        status = os.lstat(path)
        if stat.S_ISREG(status.st_mode):
            return {"bytes": status.st_size, "exact": True}
        return {"bytes": 0, "exact": False}
    except OSError:
        return {"bytes": 0, "exact": False}


def tree_size(root: str) -> dict[str, Any]:
    if not Path(root).exists():
        return {"bytes": 0, "exact": True}
    bytes_total = 0
    exact = True
    pending = [root]
    while pending:
        path = pending.pop()
        try:
            status = os.lstat(path)
            if stat.S_ISLNK(status.st_mode):
                exact = False
            elif stat.S_ISREG(status.st_mode):
                bytes_total += status.st_size
            elif stat.S_ISDIR(status.st_mode):
                pending.extend(str(Path(path) / name) for name in os.listdir(path))
            else:
                exact = False
        except OSError:
            exact = False
    return {"bytes": bytes_total, "exact": exact}


def network_usage() -> dict[str, Any]:
    if sys.platform != "linux" or not Path(NETWORK_ROOT).exists():
        return {"bytes": 0, "exact": False}
    bytes_total = 0
    observed = False
    try:
        for interface_name in os.listdir(NETWORK_ROOT):
            if interface_name == "lo":
                continue
            observed = True
            for counter in ("rx_bytes", "tx_bytes"):
                value = int(
                    Path(NETWORK_ROOT, interface_name, "statistics", counter).read_text(
                        encoding="utf-8"
                    )
                )
                if value < 0:
                    return {"bytes": 0, "exact": False}
                bytes_total += value
    except (OSError, ValueError):
        return {"bytes": 0, "exact": False}
    return {"bytes": bytes_total, "exact": observed}


def preflight() -> None:
    if sys.platform != "linux":
        refuse("linux_required")
    try:
        if not os.access(BUBBLEWRAP, os.X_OK) or not os.access(WORKSPACE, os.R_OK | os.X_OK):
            refuse("runtime_dependency_missing")
        probe = subprocess.run(
            [
                BUBBLEWRAP,
                "--die-with-parent",
                "--unshare-net",
                "--unshare-pid",
                "--unshare-uts",
                "--unshare-ipc",
                "--ro-bind",
                "/",
                "/",
                "--bind",
                WORKSPACE,
                WORKSPACE,
                "--tmpfs",
                "/run",
                "--proc",
                "/proc",
                "--dev",
                "/dev",
                "--chdir",
                WORKSPACE,
                "/bin/true",
            ],
            cwd=WORKSPACE,
            env={"PATH": "/usr/bin:/bin"},
            capture_output=True,
            text=True,
            timeout=5,
            check=False,
        )
        if probe.returncode != 0:
            refuse("bubblewrap_probe_failed")
    except (OSError, subprocess.TimeoutExpired):
        refuse("runtime_dependency_missing")
    emit(
        {
            "schema": "openagents.forensic_worker_preflight.v1",
            "driverRef": DRIVER_REF,
            "linux": True,
            "bubblewrap": True,
            "networkNamespace": "unshared",
            "workspaceRoot": "workspace",
        }
    )


def usage() -> None:
    exact = True
    tokens = 0
    active_turns = 0
    for directory in directories(TURN_ROOT, False):
        if workload_is_live_at(directory):
            active_turns += 1
        try:
            state = json.loads(Path(directory, "state.json").read_text(encoding="utf-8"))
            usage_events = [
                event
                for event in (state.get("events") or [])
                if isinstance(event, dict) and event.get("_tag") == "RuntimeUsageRecorded"
            ]
            if len(usage_events) != 1:
                exact = False
            for event in usage_events:
                usage_value = event.get("usage") or {}
                if usage_value.get("exact") is not True:
                    exact = False
                input_tokens = usage_value.get("inputTokens")
                output_tokens = usage_value.get("outputTokens")
                if not isinstance(input_tokens, int) or not isinstance(output_tokens, int):
                    exact = False
                else:
                    tokens += input_tokens + output_tokens
        except (OSError, json.JSONDecodeError, TypeError):
            exact = False
    source = tree_size(SOURCE_ROOT)
    artifact = file_size(ARTIFACT_PATH)
    network = network_usage()
    emit(
        {
            "exact": exact and source["exact"] and artifact["exact"] and network["exact"],
            "tokens": tokens,
            "sourceBytes": source["bytes"],
            "artifactBytes": artifact["bytes"],
            "networkBytes": network["bytes"],
            "activeTurns": active_turns,
        }
    )


def prepare_stop_at(
    *,
    turn_root: str,
    source_root: str,
    artifact_path: str,
    runtime_root: str,
    observe_guarded_processes: Callable[[list[str]], dict[str, Any]] = observe_guarded_processes_at,
) -> dict[str, Any]:
    guarded_roots = [turn_root, source_root, artifact_path, runtime_root]
    if any(workload_is_live_at(directory) for directory in directories(turn_root, True)):
        refuse("forensic_process_still_active")
    before = observe_guarded_processes(guarded_roots)
    if sys.platform == "linux" and not before["supported"]:
        refuse("forensic_process_observation_unavailable")
    if before["processes"]:
        refuse("forensic_process_still_active")
    shutil.rmtree(turn_root, ignore_errors=True)
    shutil.rmtree(source_root, ignore_errors=True)
    Path(artifact_path).unlink(missing_ok=True)
    shutil.rmtree(runtime_root, ignore_errors=True)
    after = observe_guarded_processes(guarded_roots)
    if sys.platform == "linux" and not after["supported"]:
        refuse("forensic_process_observation_unavailable")
    if sys.platform == "linux" and after["inaccessible"] > 0:
        refuse("forensic_process_observation_incomplete")
    scratch_paths_remaining = len([root for root in guarded_roots if path_present(root)])
    active_process_groups = len(after["processGroups"])
    if after["processes"]:
        refuse("forensic_process_still_active")
    if scratch_paths_remaining != 0:
        refuse("forensic_scratch_still_present")
    return {
        "schema": "openagents.forensic_worker_prepare_stop.v1",
        "driverRef": DRIVER_REF,
        "processObservation": (
            "proc"
            if after["supported"] and after["inaccessible"] == 0
            else "partial"
            if after["supported"]
            else "unavailable"
        ),
        "zeroProcess": len(after["processes"]) == 0 and active_process_groups == 0,
        "zeroScratch": scratch_paths_remaining == 0,
        "activeProcessGroups": active_process_groups,
        "scratchPathsRemaining": scratch_paths_remaining,
    }


def prepare_stop() -> None:
    emit(
        prepare_stop_at(
            turn_root=TURN_ROOT,
            source_root=SOURCE_ROOT,
            artifact_path=ARTIFACT_PATH,
            runtime_root=RUNTIME_ROOT,
        )
    )


def main() -> None:
    if len(sys.argv) != 2:
        refuse("unsupported_operation")
    operation = sys.argv[1]
    if operation == "preflight":
        preflight()
    elif operation == "usage":
        usage()
    elif operation == "prepare-stop":
        prepare_stop()
    else:
        refuse("unsupported_operation")


if __name__ == "__main__":
    main()
